plugin/pkg/admission/runtimeclass/admission.go

```diff
@@ -0,0 +1,199 @@
+	r.client = client
+}
+
+func (r *RuntimeClass) ValidateInitialization() error {
+	if r.lister == nil {
+		return fmt.Errorf("missing RuntimeClass lister")
+	}
+	if r.client == nil {
+		return fmt.Errorf("missing client")
+	}
+	return nil
+}
+
+// Admit makes an admission decision based on the request attributes
+func (r *RuntimeClass) Admit(attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	return admissionAction(r, attributes, "admit")
+}
+
+// Validate makes sure that pod adhere's to RuntimeClass's definition
+func (r *RuntimeClass) Validate(attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	return admissionAction(r, attributes, "validate")
+}
+
+// NewRuntimeClass creates a new always pull images admission control handler
+func NewRuntimeClass() *RuntimeClass {
+	return &RuntimeClass{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+// admissionAction handles Admit and Validate phases of admission, switching based on the admissionPhase parameter
+func admissionAction(r *RuntimeClass, attributes admission.Attributes, admissionPhase string) (err error) {
+
+	// Ignore all calls to subresources or resources other than pods.
+	if shouldIgnore(attributes) {
+		return nil
+	}
+	pod, ok := attributes.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	// Get RuntimeClass. Not finding a given runtimeClass does not return an error. We will want to continue in the
+	// case of Validate phase, even if a RuntimeClass is not defined
+	runtimeClass, err := getRuntimeClass(r, pod, pod.Spec.RuntimeClassName)
+	if err != nil {
+		return err
+	}
+
+	switch admissionPhase {
+	case "admit":
+		err := setOverhead(attributes, pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	case "validate":
+		err := validateOverhead(attributes, pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("invalid value for admissionPhase: %s", admissionPhase)
+	}
+
+	return nil
+}
+
+// getRuntimeClass will return a reference to the RuntimeClass object if it is found. If it cannot be found, or a RuntimeClassName
+// is not provided in the pod spec, *node.RuntimeClass returned will be nil
+func getRuntimeClass(r *RuntimeClass, pod *api.Pod, runtimeClassName *string) (*node.RuntimeClass, error) {
+
+	runtimeClass := &node.RuntimeClass{}
+
+	if runtimeClassName != nil {
+		v1beta1runtimeClass, _ := r.lister.Get(*runtimeClassName)
+		if v1beta1runtimeClass != nil {
+			err := nodev1beta1.Convert_v1beta1_RuntimeClass_To_node_RuntimeClass(v1beta1runtimeClass, runtimeClass, nil)
+			if err != nil {
+				return runtimeClass, err
+			}
+		}
+	}
+	return runtimeClass, nil
+}
+
+func setOverhead(a admission.Attributes, pod *api.Pod, runtimeClass *node.RuntimeClass) (err error) {
+
+	if pod.Spec.Overhead != nil {
```

Comment:
I think we should accept the overhead if it's identical to what we were going to set, since it makes the controller idempotent.

Reply:
Done.